eth_utils/curried/__init__.py

```diff
@@ -82,20 +82,122 @@
 )
 from eth_utils.toolz import curry
+from typing import Any, Callable, Sequence, Tuple, TypeVar, Union, overload
+
+TReturn = TypeVar("TReturn")
+TValue = TypeVar("TValue")
+
+
+@overload
+def apply_formatter_if(
+    condition: Callable[..., bool]
+) -> Callable[[Callable[..., TReturn]], Callable[[TValue], Union[TReturn, TValue]]]:
+    ...
+
+
+@overload
+def apply_formatter_if(
+    condition: Callable[..., bool], formatter: Callable[..., TReturn]
+) -> Callable[[TValue], Union[TReturn, TValue]]:
+    ...
+
+
+@overload
+def apply_formatter_if(
+    condition: Callable[..., bool], formatter: Callable[..., TReturn], value: TValue
+) -> Union[TReturn, TValue]:
+    ...
+
+
+# This is just a stub to appease mypy, it gets overwritten later
+def apply_formatter_if(
+    condition: Callable[..., bool],
+    formatter: Callable[..., TReturn] = None,
+    value: TValue = None,
+) -> Union[
+    Callable[[Callable[..., TReturn]], Callable[[TValue], Union[TReturn, TValue]]],
+    Callable[[TValue], Union[TReturn, TValue]],
+    TReturn,
+    TValue,
+]:
+    ...
+
+
+@overload
+def apply_one_of_formatters(
+    formatter_condition_pairs: Sequence[
+        Tuple[Callable[..., bool], Callable[..., TReturn]]
+    ]
+) -> Callable[[TValue], TReturn]:
+    ...
+
+
+@overload
+def apply_one_of_formatters(
+    formatter_condition_pairs: Sequence[
+        Tuple[Callable[..., bool], Callable[..., TReturn]]
+    ],
+    value: TValue,
+) -> TReturn:
+    ...
+
+
+# This is just a stub to appease mypy, it gets overwritten later
+def apply_one_of_formatters(
+    formatter_condition_pairs: Sequence[
+        Tuple[Callable[..., bool], Callable[..., TReturn]]
+    ],
+    value: TValue = None,
+) -> TReturn:
+    ...
+
+
+@overload
+def hexstr_if_str(
+    to_type: Callable[..., TReturn]
+) -> Callable[[Union[bytes, int, str]], TReturn]:
+    ...
+
+
+@overload
+def hexstr_if_str(
+    to_type: Callable[..., TReturn], to_format: Union[bytes, int, str]
+) -> TReturn:
+    ...
+
+
+# This is just a stub to appease mypy, it gets overwritten later
+def hexstr_if_str(
+    to_type: Callable[..., TReturn], to_format: Union[bytes, int, str] = None
+) -> TReturn:
+    ...
+
 
 apply_formatter_at_index = curry(apply_formatter_at_index)
-apply_formatter_if = curry(apply_formatter_if)
+apply_formatter_if = curry(non_curried_apply_formatter_if)
 apply_formatter_to_array = curry(apply_formatter_to_array)
 apply_formatters_to_dict = curry(apply_formatters_to_dict)
 apply_formatters_to_sequence = curry(apply_formatters_to_sequence)
 apply_key_map = curry(apply_key_map)
-apply_one_of_formatters = curry(apply_one_of_formatters)
+apply_one_of_formatters = curry(non_curried_apply_one_of_formatters)
 from_wei = curry(from_wei)
 get_logger = curry(get_logger)
-hexstr_if_str = curry(hexstr_if_str)
+hexstr_if_str = curry(non_curried_hexstr_if_str)
 is_same_address = curry(is_same_address)
 text_if_str = curry(text_if_str)
 to_wei = curry(to_wei)
 clamp = curry(clamp)
 
+del Any
```

Comment:
Curious at using this approach as opposed to `__all__`?  At minimum a comment above these `del` statements explaining why they are here would :smile:

Reply:
Didn't put too much thought into it. Basically something like:
- There was already a `del` there
- The list of things to delete is shorter than the list of things to import

I'll try it out and see how it it looks the other way.

Reply:
Hm, `__all__` doesn't actually remove the others from the import list, it only affects `from eth_utils.curried import *`, which we don't use anyway.

When I use `__all__`, this happens:
```py
>>> import eth_utils.curried
>>> eth_utils.curried.Any
typing.Any
>>> from eth_utils.curried import *
>>> to_hex
<function eth_utils.conversions.to_hex(primitive:Union[bytes, int]=None, hexstr:<function NewType.<locals>.new_type at 0x7fcf40edb158>=None, text:str=None) -> <function NewType.<locals>.new_type at 0x7fcf40edb158>>
>>> Any
NameError: name 'Any' is not defined
>>> from eth_utils.curried import Any
>>> Any
typing.Any

Reply:
Yeah, I forgot that `__all__` only effects `*`.  Dumb python...